tests/Microsoft.Identity.Test.Unit.net45/CoreTests/InstanceTests/InstanceDiscoveryManagerTests.cs

```diff
@@ -29,19 +70,122 @@ await ValidateSelfEntryAsync(new Uri(MsalTestConstants.ADFSAuthority))
+                Authority,
+                new[] { "env1", "env2" },
+                _testRequestContext)
+                .ConfigureAwait(false);
+            _staticMetadataProvider.Received(1).GetMetadata("some_env.com");
+
+            InstanceDiscoveryMetadataEntry actualResult2 = await _discoveryManager.GetMetadataEntryAsync(
+                "https://some_env.com/tid",
+                _testRequestContext)
+                .ConfigureAwait(false);
+            _staticMetadataProvider.Received(2).GetMetadata("some_env.com");
+            _staticMetadataProvider.AddMetadata(null, null);
+
+            // Assert
+            Assert.AreSame(_expectedResult, actualResult1, "The static provider should be queried first");
+            Assert.AreSame(_expectedResult, actualResult2, "The static provider should be queried first");
+        }
+
+        [TestMethod]
+        public async Task KnownMetadataProviderIsCheckedSecondAsync()
+        {
+            // Arrange
+            var otherEnvs = new[] { "env1", "env2" };
+
+            // No response from the static provider
+            _staticMetadataProvider.GetMetadata("some_env.com").Returns((InstanceDiscoveryMetadataEntry)null);
+            _knownMetadataProvider.GetMetadata("some_env.com", otherEnvs).Returns(_expectedResult);
+
+            // Act
+            InstanceDiscoveryMetadataEntry actualResult = await _discoveryManager.GetMetadataEntryTryAvoidNetworkAsync(
+                "https://some_env.com/tid",
+                otherEnvs,
+                _testRequestContext)
+                .ConfigureAwait(false);
+
+            // Assert
+            Assert.AreSame(_expectedResult, actualResult, "The known metadata provider should be queried second");
+            _staticMetadataProvider.Received(1).GetMetadata("some_env.com");
+            _knownMetadataProvider.Received(1).GetMetadata("some_env.com", otherEnvs);
+        }
+
+        [TestMethod]
+        public async Task NetworkProviderIsCalledLastAsync()
+        {
+            // Arrange
+            _staticMetadataProvider = new StaticMetadataProvider();
+
+            _discoveryManager = new InstanceDiscoveryManager(
+                _harness.HttpManager,
+                _harness.ServiceBundle.TelemetryManager,
+                false,
+                _knownMetadataProvider,
+                _staticMetadataProvider,
+                _networkMetadataProvider);
+
+            var otherEnvs = new[] { "env1", "env2" };
+            InstanceDiscoveryResponse discoveryResponse = new InstanceDiscoveryResponse
+            {
+                Metadata = new[] { _expectedResult }
+            };
+            var authorityUri = new Uri(Authority);
+
+            // No response from the static and known provider
+            _knownMetadataProvider
+                .GetMetadata("some_env.com", otherEnvs)
+                .Returns((InstanceDiscoveryMetadataEntry)null);
```

Comment:
do you need to cast this since it's null?

Reply:
Yeah, there are lots of Return() overloads and .net complains that it can't choose the right one if I just give it null.